<div align="center">

# 📊 S13 — NumPy aplicado al análisis de portafolios
### Lenguajes de Programación · Ingeniería Financiera

*De la lista al array — y del array a una recomendación de inversión*

</div>

---

En el reto del portafolio (S11-S12) escribiste tus propias funciones. Hoy conoces NumPy — y lo vas a usar no solo para calcular, sino para **analizar y justificar decisiones**, como lo haría un analista de verdad.

| Bloque | Tema |
|:---:|---|
| 00 | ¿Qué es una librería? |
| 01 | Qué es un array — y cuántas dimensiones puede tener |
| 02 | Creación de arrays |
| 03 | Operaciones básicas |
| 04 | Analiza un activo: rendimiento y riesgo en el tiempo |
| 05 | Compara los 4 activos |
| 06 | Construye un portafolio ponderado |
| 07 | 🎯 Reto individual |

## 00 · ¿Qué es una librería?

> **Nota** — Objetivo: entender qué es una librería antes de usar una.

Una librería es un paquete de funciones que otra persona ya escribió, probó y publicó para que cualquiera las use — como las tuyas, pero construidas para casos generales. **NumPy** es una librería con funciones ya hechas para trabajar con muchos números a la vez.

**¿Cuál es el objetivo de NumPy?** Darle a Python la capacidad de hacer cálculos numéricos rápido sobre conjuntos grandes de datos.

In [1]:
import numpy as np

print("NumPy", np.__version__, "listo para usar.")

NumPy 2.4.4 listo para usar.


> 💡 **Idea clave** — `import` le dice a Python que traiga la librería. `numpy` es su nombre. `as np` le da un apodo corto — es la convención que usa toda la comunidad.

---
## 01 · Qué es un array — y cuántas dimensiones puede tener

> **Nota** — Objetivo: entender qué es un array, y que puede tener una o más dimensiones.

Un array es una colección de números, todos del mismo tipo, donde cada uno tiene una posición fija — su **índice**.

In [2]:
precios = np.array([195, 385, 185, 165])   # AAPL, MSFT, AMZN, GOOGL

print(precios)
print("shape:", precios.shape)
print("dtype:", precios.dtype)

[195 385 185 165]
shape: (4,)
dtype: int64


Esta fila tiene **una sola dimensión** — un **eje (axis)**. Pero un array puede tener dos ejes: una tabla, con filas y columnas. Por ejemplo, el precio de 4 activos durante 5 días:

| | Día 1 | Día 2 | Día 3 | Día 4 | Día 5 |
|---|---|---|---|---|---|
| **AAPL** | 192 | 194 | 193 | 196 | 195 |
| **MSFT** | 390 | 388 | 386 | 384 | 385 |

In [3]:
precios_5_dias = np.array([
    [192, 194, 193, 196, 195],   # AAPL
    [390, 388, 386, 384, 385],   # MSFT
])

print(precios_5_dias)
print("shape:", precios_5_dias.shape, "  # (2 activos, 5 días)")

[[192 194 193 196 195]
 [390 388 386 384 385]]
shape: (2, 5)   # (2 activos, 5 días)


> 💡 **Idea clave** — Más adelante en esta clase, cada activo va a tener justo esta forma: **una fila de rendimientos a través de varios días**. Guarda esta idea — es la base de todo lo que sigue.

---
## 02 · Creación de arrays

> **Nota** — Objetivo: crear un array a partir de una lista.

**✏️ Completa el código**

In [4]:
cantidades = np.array([10, 8, 12, 6])   # 10, 8, 12, 6
rendimientos = np.array([8.33, -3.75, 8.82, 10.00]) # 8.33, -3.75, 8.82, 10.00

print(cantidades, cantidades.dtype)
print(rendimientos, rendimientos.dtype)

[10  8 12  6] int64
[ 8.33 -3.75  8.82 10.  ] float64


<details>
<summary>💡 Ver respuesta</summary>

```python
cantidades = np.array([10, 8, 12, 6])
rendimientos = np.array([8.33, -3.75, 8.82, 10.00])
```
</details>

> ⚠️ **Error frecuente** — `np.array(1, 2, 3, 4)` falla: los números deben ir dentro de **una sola lista**. Pruébalo (esto debe fallar):

In [5]:
np.array(1, 2, 3, 4)

TypeError: array() takes from 1 to 2 positional arguments but 4 were given

---
## 03 · Operaciones básicas

> **Nota** — Objetivo: operar el array completo, sin ciclos.

Los operadores aritméticos se aplican **elemento por elemento**.

In [6]:
precios_actuales = np.array([195, 385, 185, 165])
precios_compra = np.array([180, 400, 170, 150])

valor_invertido = cantidades * precios_compra
valor_actual = cantidades * precios_actuales

print("Valor actual:", valor_actual)

Valor actual: [1950 3080 2220  990]


**✏️ Completa el código** — calcula `ganancia` y `en_perdida` (verdadero/falso).

In [7]:
ganancia = valor_actual - valor_invertido
en_perdida = ganancia < 0

print("Ganancia:", ganancia)
print("En pérdida:", en_perdida)

Ganancia: [ 150 -120  180   90]
En pérdida: [False  True False False]


<details>
<summary>💡 Ver respuesta</summary>

```python
ganancia = valor_actual - valor_invertido
en_perdida = ganancia < 0
```
</details>

> 💡 **Idea clave** — Ese array de `True`/`False` es exactamente lo que se usa por dentro para **filtrar** una tabla en pandas (S14).

---
## 04 · Analiza un activo: rendimiento y riesgo en el tiempo

> **Nota** — Objetivo: usar `mean()` y `std()` para describir el comportamiento de un activo a través de varios días — no solo calcularlos.

A partir de aquí trabajas con algo distinto: no es "un número por activo", sino **la serie de rendimientos diarios de AAPL durante 8 días** de negociación.

In [8]:
AAPL = np.array([0.012, 0.008, -0.004, 0.015, 0.006, -0.002, 0.010, 0.003])

print("Número de días:", len(AAPL))
print("Primer día:", AAPL[0])
print("Último día:", AAPL[-1])

Número de días: 8
Primer día: 0.012
Último día: 0.003


**🔎 Pregunta de análisis**

Un compañero mira el promedio de AAPL y afirma: *"Como el rendimiento promedio es positivo, quiere decir que AAPL tuvo un buen día todos los días."*

¿Tiene razón? Usa el propio array `AAPL` (no solo el promedio) para comprobarlo antes de responder.

In [9]:
promedio_aapl = AAPL.mean()
peor_dia = AAPL.min()

print("Promedio:", promedio_aapl)
print("Peor día:", peor_dia)
print("Días negativos:", (AAPL < 0).sum())

Promedio: 0.006
Peor día: -0.004
Días negativos: 2


**✍️ Mi respuesta**

No tiene razón. El promedio de AAPL es positivo (+0.60 %), pero el array muestra **2 días negativos** (-0.4 % y -0.2 %): `AAPL.min()` da -0.4 % y `(AAPL < 0).sum()` da 2. El promedio resume la tendencia de los 8 días; no dice que cada día individual haya sido bueno. Para afirmar "todos los días fueron buenos" hay que revisar la serie completa, no solo su resumen.


<details>
<summary>💡 Ver una forma de responder</summary>

No tiene razón. El promedio (0.6%) es positivo, pero el peor día fue **-0.4%** — hubo al menos un día con pérdida. El promedio resume la tendencia general, no garantiza que cada día individual haya sido positivo. Para saber eso hay que mirar el array completo, no solo su resumen.
</details>

**✏️ Ahora tú:** calcula el riesgo de AAPL (su desviación estándar).

In [10]:
riesgo_aapl = AAPL.std()
print("Riesgo (std):", riesgo_aapl)

Riesgo (std): 0.006224949798994366


<details>
<summary>💡 Ver respuesta</summary>

```python
riesgo_aapl = AAPL.std()
```
`std()` ≈ 0.0062 (0.62%).
</details>

Ese número mide qué tanto se aleja cada día del promedio, en ambas direcciones — no cuánto ganas o pierdes en un día cualquiera, sino qué tan predecible o errático fue el activo durante esos 8 días. Por eso a la desviación estándar se le trata, en la práctica, como sinónimo de **volatilidad**: dos activos pueden tener el mismo rendimiento promedio y aun así uno ser mucho más riesgoso que el otro, simplemente porque sus días buenos y malos son más extremos.

---
## 05 · Compara los 4 activos

> **Nota** — Objetivo: comparar rendimiento y riesgo de varios activos a la vez, y tomar una posición justificada — no solo calcular cada uno por separado.

Estas son las series de 8 días de tus 4 activos:

In [11]:
AAPL = np.array([0.012, 0.008, -0.004, 0.015, 0.006, -0.002, 0.010, 0.003])
MSFT = np.array([0.004, 0.003, 0.005, 0.002, 0.006, 0.004, 0.003, 0.005])
AMZN = np.array([-0.010, -0.015, 0.002, -0.008, -0.012, 0.005, -0.020, -0.006])
TSLA = np.array([0.030, -0.025, 0.040, 0.015, -0.020, 0.035, -0.010, 0.020])

**✏️ Completa la tabla** — calcula el rendimiento promedio y el riesgo de cada uno.

In [12]:
activos = {"AAPL": AAPL, "MSFT": MSFT, "AMZN": AMZN, "TSLA": TSLA}

for nombre, serie in activos.items():
    promedio = serie.mean()
    riesgo = serie.std()
    print(f"{nombre}: rendimiento {promedio:.4f}  |  riesgo {riesgo:.4f}")

AAPL: rendimiento 0.0060  |  riesgo 0.0062
MSFT: rendimiento 0.0040  |  riesgo 0.0012
AMZN: rendimiento -0.0080  |  riesgo 0.0078
TSLA: rendimiento 0.0106  |  riesgo 0.0239


<details>
<summary>💡 Ver respuesta</summary>

```python
for nombre, serie in activos.items():
    promedio = serie.mean()
    riesgo = serie.std()
    print(f"{nombre}: rendimiento {promedio:.4f}  |  riesgo {riesgo:.4f}")
```

| Activo | Rendimiento promedio | Riesgo (std) |
|---|---|---|
| AAPL | 0.60% | 0.62% |
| MSFT | 0.40% | 0.12% |
| AMZN | -0.80% | 0.78% |
| TSLA | 1.06% | 2.39% |
</details>

**🔎 Preguntas de análisis** — respóndelas en una celda de texto, con la tabla de arriba como evidencia:

1. Si **solo** te importara maximizar el rendimiento promedio, ¿qué activo elegirías? Si **solo** te importara minimizar el riesgo, ¿cuál elegirías? ¿Es el mismo activo en los dos casos? ¿Qué te dice eso?
2. Un cliente te dice: *"Nunca invertiría en algo con tanto riesgo como TSLA."* Otro cliente te dice: *"Nunca invertiría en algo que en promedio pierde dinero, como AMZN."* ¿Estás de acuerdo con ambos? Justifica usando los números de la tabla, no solo tu opinión.
3. Recomienda **un solo activo** para un cliente que prioriza la estabilidad por encima de todo. Justifica tu elección considerando rendimiento *y* riesgo — no solo uno de los dos.

**✍️ Mis respuestas (sección 05)**

**1.** Para maximizar el rendimiento promedio elegiría **TSLA** (1.06 %); para minimizar el riesgo elegiría **MSFT** (0.12 %). No es el mismo activo. Eso muestra que, en esta muestra, el activo que más rinde también es el más volátil (TSLA tiene un riesgo de 2.39 %) y el más estable rinde menos (MSFT, 0.40 %): normalmente hay que aceptar más riesgo para aspirar a más rendimiento, y la elección depende del perfil del inversionista.

**2.** *Cliente de TSLA:* de acuerdo solo en parte. Es cierto que TSLA es el más riesgoso (2.39 %, casi 4 veces el de AAPL y unas 20 veces el de MSFT), pero también es el de mayor rendimiento promedio (1.06 %); decir "nunca" ignora la mitad de la ecuación y no considera cuánto peso tendría en el portafolio. *Cliente de AMZN:* de acuerdo. Es el único con promedio negativo (-0.80 %), perdió en 6 de los 8 días y su riesgo (0.78 %) es mayor que el de AAPL y MSFT: peor rendimiento sin compensarlo con estabilidad. La salvedad es que son solo 8 días de datos.

**3.** Recomendaría **MSFT**: tiene rendimiento promedio positivo (+0.40 %), el riesgo más bajo (0.12 %) y ningún día negativo (su peor día fue +0.2 %). AAPL rinde más (0.60 %), pero con un riesgo unas 5 veces mayor (0.62 %), que no se justifica para quien prioriza la estabilidad.


---
## 06 · Construye un portafolio ponderado

> **Nota** — Objetivo: combinar varios activos con distintos pesos, y entender qué le pasa al riesgo cuando diversificas.

Un portafolio real casi nunca es "todo en un solo activo" — se reparte el dinero entre varios, con distintos pesos. Supón esta distribución:

In [13]:
pesos = np.array([0.35, 0.40, 0.10, 0.15])  # AAPL, MSFT, AMZN, TSLA — suman 1.0

rendimiento_portafolio = AAPL*pesos[0] + MSFT*pesos[1] + AMZN*pesos[2] + TSLA*pesos[3]
print(rendimiento_portafolio)

[ 0.0093  -0.00125  0.0068   0.0075   0.0003   0.00665  0.0012   0.00545]


> 💡 **Idea clave** — Una sola operación combinó los 4 activos, día por día, sin ningún `for`. Cada día del portafolio es el promedio ponderado de lo que hicieron los 4 activos ese día.

**✏️ Completa el código** — calcula el rendimiento promedio y el riesgo del portafolio.

In [14]:
rendimiento_promedio_portafolio = rendimiento_portafolio.mean()
riesgo_portafolio = rendimiento_portafolio.std()

print(f"Rendimiento promedio del portafolio: {rendimiento_promedio_portafolio:.4f}")
print(f"Riesgo del portafolio: {riesgo_portafolio:.4f}")

Rendimiento promedio del portafolio: 0.0045
Riesgo del portafolio: 0.0036


<details>
<summary>💡 Ver respuesta</summary>

```python
rendimiento_promedio_portafolio = rendimiento_portafolio.mean()
riesgo_portafolio = rendimiento_portafolio.std()
```

Rendimiento promedio ≈ **0.45%**, riesgo ≈ **0.36%**.
</details>

**🔎 Pregunta de análisis**

Compara el riesgo del portafolio (≈0.36%) con el promedio ponderado de los riesgos *individuales* de los 4 activos, calculado abajo:

In [15]:
riesgos_individuales = np.array([AAPL.std(), MSFT.std(), AMZN.std(), TSLA.std()])
riesgo_promedio_simple = (riesgos_individuales * pesos).sum()

print(f"Riesgo si solo promediaras los riesgos individuales: {riesgo_promedio_simple:.4f}")
print(f"Riesgo real del portafolio combinado: {riesgo_portafolio:.4f}")

Riesgo si solo promediaras los riesgos individuales: 0.0070
Riesgo real del portafolio combinado: 0.0036


El riesgo real del portafolio (≈0.36%) es **casi la mitad** de lo que esperarías si solo promediaras el riesgo de cada activo por separado (≈0.70%). Ningún activo individual cambió de riesgo — el portafolio completo sí.

**Explica con tus palabras**, en una celda de texto, por qué combinar activos distintos puede reducir el riesgo total sin que ninguno de ellos se vuelva menos riesgoso por su cuenta. (Pista: fíjate en qué días TSLA pierde y qué hacen AAPL o MSFT esos mismos días.)

In [16]:
# Evidencia: ¿qué hacen AAPL y MSFT los días en que TSLA pierde?
dias_tsla_pierde = TSLA < 0
print("TSLA pierde los días:", np.where(dias_tsla_pierde)[0] + 1)
print("AAPL esos días:", AAPL[dias_tsla_pierde])
print("MSFT esos días:", MSFT[dias_tsla_pierde])
print("Correlación TSLA-AAPL:", np.corrcoef(TSLA, AAPL)[0, 1].round(2))

TSLA pierde los días: [2 5 7]
AAPL esos días: [0.008 0.006 0.01 ]
MSFT esos días: [0.003 0.006 0.003]
Correlación TSLA-AAPL: -0.43


**✍️ Mi explicación (sección 06)**

Combinar activos reduce el riesgo porque no se mueven igual los mismos días. TSLA pierde los días 2, 5 y 7, y justo esos días AAPL gana (+0.8 %, +0.6 % y +1.0 %) y MSFT también (+0.3 %, +0.6 % y +0.3 %), así que sus ganancias amortiguan la caída de TSLA. La correlación entre TSLA y AAPL es negativa (≈ -0.43): tienden a moverse en direcciones opuestas. Por eso el riesgo del portafolio (≈ 0.36 %) es casi la mitad del promedio ponderado de los riesgos individuales (≈ 0.70 %), aunque ningún activo haya cambiado su propio riesgo: ese efecto es la **diversificación**.


<details>
<summary>💡 Ver una forma de responder</summary>

Porque los activos no suben y bajan exactamente el mismo día ni en la misma dirección. Cuando uno pierde, otro puede sostenerse o ganar, y eso amortigua el resultado combinado — es el efecto de la **diversificación**.

Dicho de otra forma: el riesgo de un portafolio no depende solo de cuánto se mueve cada activo por separado, sino de qué tan parecido se mueven entre sí — su **correlación**. Cuando dos activos tienden a moverse en direcciones opuestas (uno sube cuando el otro baja), combinarlos reduce el riesgo total mucho más que si simplemente promediaras sus riesgos individuales. Es la razón por la que un portafolio bien armado casi siempre mezcla activos que no reaccionan igual a las mismas condiciones del mercado.
</details>

**✏️ Pruébalo tú mismo** — cambia los pesos abajo (deben sumar 1.0) y observa qué le pasa al riesgo del portafolio.

In [17]:
mis_pesos = np.array([0.25, 0.25, 0.25, 0.25])  # cambia estos números — deben sumar 1.0

mi_portafolio = AAPL*mis_pesos[0] + MSFT*mis_pesos[1] + AMZN*mis_pesos[2] + TSLA*mis_pesos[3]
print(f"Rendimiento: {mi_portafolio.mean():.4f}  |  Riesgo: {mi_portafolio.std():.4f}")

Rendimiento: 0.0032  |  Riesgo: 0.0070


---
## 🎯 07 · Reto individual

Retoma el portafolio que te asignaron en el reto de S11-S12.

In [18]:
# Datos asignados a mí (los de "Copiar mis datos para el cuaderno").
# Misma estructura que el diccionario de la plantilla: cantidad y precio_compra por activo.
mi_portafolio = {
    "NFLX": {"cantidad": 16, "precio_compra": 284},
    "V":    {"cantidad": 14, "precio_compra": 295},
    "TSLA": {"cantidad": 15, "precio_compra": 423},
    "JPM":  {"cantidad": 8,  "precio_compra": 288},
}
mis_precios = {"NFLX": 298, "V": 306, "TSLA": 352, "JPM": 232}

print("Mi portafolio asignado:")
print(mi_portafolio)
print("Precios actuales:", mis_precios)


Mi portafolio asignado:
{'NFLX': {'cantidad': 16, 'precio_compra': 284}, 'V': {'cantidad': 14, 'precio_compra': 295}, 'TSLA': {'cantidad': 15, 'precio_compra': 423}, 'JPM': {'cantidad': 8, 'precio_compra': 288}}
Precios actuales: {'NFLX': 298, 'V': 306, 'TSLA': 352, 'JPM': 232}


### Tu tarea

Con `mi_portafolio` y `mis_precios`, usando **solo arrays de NumPy**:

1. Calcula `valor_invertido`, `valor_actual`, `ganancia` y `rendimiento` (%) de tus 4 activos.
2. Calcula el promedio, el mejor, el peor y la desviación estándar del rendimiento.
3. **Escribe una recomendación de una sola frase** para tu portafolio, dirigida a un inversionista, que considere tanto el rendimiento como el riesgo — no solo uno de los dos. Debe quedar claro, leyendo tu frase, en qué números te basaste.

In [19]:
import numpy as np

# --- 1) Arrays (mismo orden de activos en los tres) ---
activos = list(mi_portafolio.keys())
cantidades       = np.array([mi_portafolio[a]["cantidad"] for a in activos])
precios_compra   = np.array([mi_portafolio[a]["precio_compra"] for a in activos])
precios_actuales = np.array([mis_precios[a] for a in activos])

valor_invertido = cantidades * precios_compra
valor_actual    = cantidades * precios_actuales
ganancia        = valor_actual - valor_invertido
rendimiento     = ganancia / valor_invertido * 100   # en %

print(f"{'Activo':<6}{'Invertido':>10}{'Actual':>9}{'Ganancia':>10}{'Rend. %':>9}")
for a, vi, va, g, r in zip(activos, valor_invertido, valor_actual, ganancia, rendimiento):
    print(f"{a:<6}{vi:>10}{va:>9}{g:>10}{r:>9.2f}")

# --- 2) Estadísticas del rendimiento ---
promedio     = rendimiento.mean()
mejor        = rendimiento.max()
peor         = rendimiento.min()
riesgo       = rendimiento.std()          # desviación estándar (ddof=0, por defecto en NumPy)
mejor_activo = activos[rendimiento.argmax()]
peor_activo  = activos[rendimiento.argmin()]

print(f"\nPromedio: {promedio:.2f} %")
print(f"Mejor:    {mejor:.2f} % ({mejor_activo})")
print(f"Peor:     {peor:.2f} % ({peor_activo})")
print(f"Desviación estándar: {riesgo:.2f}")

# Dato adicional: rendimiento total del portafolio ponderado por lo invertido
rend_total = ganancia.sum() / valor_invertido.sum() * 100
print(f"\nRendimiento total ponderado: {rend_total:.2f} %  (ganancia total: {ganancia.sum()})")

# --- 3) Recomendación (una sola frase) ---
perdedores = " y ".join(a for a, r in zip(activos, rendimiento) if r < 0)
recomendacion = (
    f"Recomiendo no aumentar la inversión y reevaluar las posiciones en {perdedores}, "
    f"porque el rendimiento promedio es {promedio:.2f} % con una desviación estándar de {riesgo:.2f}: "
    f"un resultado negativo y muy disperso, con un rango de {peor:.2f} % a {mejor:.2f} %."
)
print("\n" + recomendacion)


Activo Invertido   Actual  Ganancia  Rend. %
NFLX        4544     4768       224     4.93
V           4130     4284       154     3.73
TSLA        6345     5280     -1065   -16.78
JPM         2304     1856      -448   -19.44

Promedio: -6.89 %
Mejor:    4.93 % (NFLX)
Peor:     -19.44 % (JPM)
Desviación estándar: 11.27

Rendimiento total ponderado: -6.55 %  (ganancia total: -1135)

Recomiendo no aumentar la inversión y reevaluar las posiciones en TSLA y JPM, porque el rendimiento promedio es -6.89 % con una desviación estándar de 11.27: un resultado negativo y muy disperso, con un rango de -19.44 % a 4.93 %.


### Exit ticket

«El rendimiento promedio de mi portafolio es **-6.89** %. La desviación estándar es **11.27**. Mi recomendación para un inversionista es no aumentar la inversión y reevaluar las posiciones en TSLA y JPM, porque el promedio es negativo y la dispersión es alta: TSLA (-16.78 %) y JPM (-19.44 %) pierden más que lo que ganan NFLX (+4.93 %) y V (+3.73 %), y el rango entre el mejor y el peor activo es de unos 24 puntos.»

---

📚 **Fuentes** · [NumPy quickstart](https://numpy.org/doc/stable/user/quickstart.html) · [What is NumPy?](https://numpy.org/doc/stable/user/whatisnumpy.html) · ["La desviación estándar: cálculo e interpretación para invertir", inbestMe](https://www.inbestme.com/es/es/blog/desviacion-estandar-volatilidad-riesgo/) · ["¿Qué significa la desviación estándar respecto a los riesgos de inversión?", Cuida tu Dinero](https://www.cuidatudinero.com/13168060/que-significa-la-desviacion-estandar-respecto-a-los-riesgos-de-inversion)